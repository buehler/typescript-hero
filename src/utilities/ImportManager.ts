import {Logger} from './Logger';
import {Configuration} from './Configuration';
import {SymbolType, TypescriptSymbol} from '../models/TypescriptSymbol';
import {ImportSymbol} from '../models/ImportSymbol';
import {SymbolCache} from './SymbolCache';
import * as vscode from 'vscode';
import * as path from 'path';

const reservedWords = ['window', 'dom', 'array', 'from', 'null', 'return', 'get', 'set', 'boolean', 'string', 'if', 'var', 'let', 'const', 'for', 'public', 'class', 'interface', 'new', 'import', 'as', 'private', 'while', 'case', 'switch', 'this', 'function', 'enum'],
    utilityKeywords = ['cancel', 'build', 'finish', 'merge', 'clamp', 'construct', 'native', 'clear', 'update', 'parse', 'sanitize', 'render', 'has', 'equal', 'dispose', 'create', 'as', 'is', 'init', 'process', 'get', 'set'],
    matchers = {
        imports: /import[\s]+[\*\{]*[\s]*([a-zA-Z\_\,\s]*)[\s]*[\}]*[\s]*from[\s]*[\'\"]([\S]*)[\'|\"]+/,
        types: /([.?_:\'\"a-zA-Z0-9]{2,})/g
    };

class ImportList {
    private imports: { [libName: string]: ImportSymbol[] } = {};

    public static create(editor: vscode.TextEditor, cache: SymbolCache): ImportList {
        let obj = new ImportList(editor, cache);

        let filename = path.parse(editor.document.fileName).name;
        for (let lineNr = 0; lineNr < editor.document.lineCount; lineNr++) {
            let line = editor.document.lineAt(lineNr);
            obj.matchImports(line.text);
            obj.matchTypes(line, filename);
        }

        return obj;
    }

    private static isValidName(name: string): boolean {
        return reservedWords.indexOf(name) === -1 && !utilityKeywords.some(o => name.startsWith(o));
    }

    constructor(private editor: vscode.TextEditor, private cache: SymbolCache) { }

    public addImport(newImport: ImportSymbol): void {
        if (newImport.symbol.type === SymbolType.Typings && this.imports[newImport.symbol.alias]) {
            return;
        }
        if (!this.imports[newImport.symbol.library]) {
            this.imports[newImport.symbol.library] = [newImport];
        } else if (!this.imports[newImport.symbol.library].some(o => o.element === newImport.element)) {
            this.imports[newImport.symbol.library].push(newImport);
        }
    }

    public commit(): void {
        this.editor.edit(builder => {
            let nonImportLine;
            for (let lineNr = 0; lineNr < this.editor.document.lineCount; lineNr++) {
                let line = this.editor.document.lineAt(lineNr);
                if (!line.text.match(matchers.imports)) {
                    nonImportLine = lineNr;
                    break;
                }
            }
            builder.replace(new vscode.Range(new vscode.Position(0, 0), new vscode.Position(nonImportLine, 0)), this.buildImports());
        });
    }

    private matchImports(lineText: string): void {
        let matches = lineText.match(matchers.imports);
        if (!matches) {
            return;
        }
        let libName = path.parse(matches[2]).name;

        let symbol = this.cache.symbolCache.find(o => o.library === libName);;

        if (!symbol) {
            Logger.instance.warn('A symbol was not found in the cache.', { symbol: libName });
            return;
            // TODO: is this needed?
            // symbol = new TypescriptSymbol(libName, '', SymbolType.Typings);
            // this.imports[libName].push(new ImportSymbol(symbol.alias, symbol));
        }

        if (!this.imports[libName]) {
            this.imports[libName] = [];
        }

        if (symbol.type === SymbolType.Typings) {
            this.imports[libName].push(new ImportSymbol(symbol.alias, symbol));
        } else {
            matches[1]
                .split(',')
                .map(o => o.trim())
                .forEach(o => this.imports[libName].push(new ImportSymbol(o, symbol)));
        }
    }

    private matchTypes(line: vscode.TextLine, filename: string): void {
        if (line.isEmptyOrWhitespace || line.text.match(matchers.imports)) {
            return;
        }
        let text = line.text.trim();
        if (text.startsWith('//') || text.startsWith('/*') || text.startsWith('*')) {
            return;
        }
        if (text.indexOf('//') !== -1) {
            text = text.split('//')[0];
        }
        let matches = text.match(matchers.types);
        if (!matches) {
            return;
        }

        for (let word of matches) {
            // only process unquoted words which are not listed in the commonList
            if (word.indexOf(`'`) > -1 || word.indexOf(`"`) > -1 || !ImportList.isValidName(word)) {
                continue;
            }

            let splitted = word.split('.');
            word = splitted[0];

            this.cache.symbolCache
                .filter(o => {
                    return o.library !== filename &&
                        (
                            (o.type === SymbolType.Typings && word === o.alias) ||
                            ((o.type === SymbolType.Local || o.type === SymbolType.Node) && o.exports.indexOf(word) > -1)
                        );
                })
                .forEach(o => {
                    if (o.type === SymbolType.Typings) {
                        if (!this.imports[o.library]) {
                            this.imports[o.library] = [new ImportSymbol(o.alias, o)];
                        }
                    } else {
                        if (this.imports[o.library] && !this.imports[o.library].some(i => i.element === word)) {
                            this.imports[o.library].push(new ImportSymbol(word, o));
                        }
                    }
                });
        }
    }

    private buildImports(): string {
        let imports = '';
        for (let lib in this.imports) {
            let symbols = this.imports[lib],
                libSymbol = symbols[0].symbol;

            if (libSymbol.type === SymbolType.Typings) {
                imports += `import * as ${libSymbol.alias} from ${Configuration.pathStringDelimiter}${libSymbol.library}${Configuration.pathStringDelimiter};\n`;
            } else if (libSymbol.type === SymbolType.Node) {
                imports += `import {${symbols.map(o => o.element).join(', ')}} from ${Configuration.pathStringDelimiter}${libSymbol.library}${Configuration.pathStringDelimiter};\n`;
            } else {
                let importPath = './' + path.relative(path.parse(this.editor.document.fileName).dir, `${libSymbol.path}/${libSymbol.library}`);
                imports += `import {${symbols.map(o => o.element).join(', ')}} from ${Configuration.pathStringDelimiter}${importPath}${Configuration.pathStringDelimiter};\n`;
            }
        }
        return imports;
    }
}

export class ImportManager {
    constructor(private cache: SymbolCache) { }

    public organizeImports(item?: ImportSymbol): void {
        let list = ImportList.create(vscode.window.activeTextEditor, this.cache);
        if (item) {
            list.addImport(item);
        }
        list.commit();
    }

    public addImport(item: ImportSymbol): void {
        this.organizeImports(item);
    }
}
