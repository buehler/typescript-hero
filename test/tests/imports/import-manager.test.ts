import { join } from 'path';
import { ClassDeclaration, DefaultDeclaration, File, ModuleDeclaration } from 'typescript-parser';
import { Position, Range, TextDocument, Uri, window, workspace } from 'vscode';

import ImportManager from '../../../src/imports/import-manager';
import ioc from '../../../src/ioc';
import iocSymbols, { ImportManagerProvider } from '../../../src/ioc-symbols';
import { expect, getDocumentText } from '../setup';

describe('ImportManager', () => {

  describe('Typescript', () => {

    const rootPath = workspace.workspaceFolders![0].uri.fsPath;
    const file = Uri.file(join(rootPath, 'imports', 'import-manager-file.ts'));
    const provider: ImportManagerProvider = ioc.get<ImportManagerProvider>(iocSymbols.importManager);
    let document: TextDocument;
    let documentText: string;

    before(async () => {
      document = await workspace.openTextDocument(file);
      await window.showTextDocument(document);

      documentText = document.getText();
    });

    afterEach(async () => {
      await window.activeTextEditor!.edit((builder) => {
        builder.delete(new Range(
          new Position(0, 0),
          document.lineAt(document.lineCount - 1).rangeIncludingLineBreak.end,
        ));
        builder.insert(new Position(0, 0), documentText);
      });

    });

    describe('Constructor (via provider)', () => {

      it('should create a document controller', async () => {
        const ctrl = await provider(document);

        expect(ctrl).to.be.an.instanceof(ImportManager);
      });

      it('should parse the document', async () => {
        const ctrl = await provider(document);

        expect((ctrl as any).parsedDocument).to.be.an.instanceof(File);
      });

      it('should add an import proxy for a named import', async () => {
        const ctrl = await provider(document);
        const imps = (ctrl as any).parsedDocument.imports;

        expect(imps).to.matchSnapshot();
      });

      it('should add an import proxy for a default import', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.replace(
            new Range(
              new Position(0, 0),
              new Position(1, 0),
            ),
            `import myDefaultExportedFunction from '../defaultExport/lateDefaultExportedElement';\n`,
          );
        });

        const ctrl = await provider(document);
        const imps = (ctrl as any).parsedDocument.imports;

        expect(imps).to.matchSnapshot();
      });

      it('should add multiple import proxies', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.insert(
            new Position(0, 0),
            `import myDefaultExportedFunction from '../defaultExport/lateDefaultExportedElement';\n`,
          );
        });

        const ctrl = await provider(document);
        const imps = (ctrl as any).parsedDocument.imports;

        expect(imps).to.matchSnapshot();
      });

      it('should not add a proxy for a namespace import', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.replace(
            new Range(
              new Position(0, 0),
              new Position(1, 0),
            ),
            `import * as bodyParser from 'body-parser';\n`,
          );
        });

        const ctrl = await provider(document);
        const imps = (ctrl as any).parsedDocument.imports;

        expect(imps).to.matchSnapshot();
      });

      it('should not add a proxy for an external import', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.replace(
            new Range(
              new Position(0, 0),
              new Position(1, 0),
            ),
            `import bodyParser = require('body-parser');\n`,
          );
        });

        const ctrl = await provider(document);
        const imps = (ctrl as any).parsedDocument.imports;

        expect(imps).to.matchSnapshot();
      });

      it('should not add a proxy for a string import', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.replace(
            new Range(
              new Position(0, 0),
              new Position(1, 0),
            ),
            `import 'body-parser';\n`,
          );
        });

        const ctrl = await provider(document);
        const imps = (ctrl as any).parsedDocument.imports;

        expect(imps).to.matchSnapshot();
      });

    });

    describe('addDeclarationImport()', () => {

      it('should add a normal import to the document', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('classdeclaration', true);
        ctrl.addDeclarationImport({ declaration, from: '../lib' });

        expect((ctrl as any).imports).to.matchSnapshot();
      });

      it('should add a module import to the import index', async () => {
        const ctrl = await provider(document);
        const declaration = new ModuleDeclaration('module');
        ctrl.addDeclarationImport({ declaration, from: 'my-module' });

        expect((ctrl as any).imports).to.matchSnapshot();
      });

      it('should add a default import to the import index.', async () => {
        const ctrl = await provider(document);

        const declaration = new DefaultDeclaration('defaultdeclaration', new File('path', rootPath, 0, 1));
        ctrl.addDeclarationImport({ declaration, from: '../lib' });

        expect((ctrl as any).imports).to.matchSnapshot();
      });

      it('should add multiple imports to the import index', async () => {
        const ctrl = await provider(document);
        const declarations = [
          new ClassDeclaration('class1', true),
          new ClassDeclaration('class2', true),
        ];
        ctrl
          .addDeclarationImport({ declaration: declarations[0], from: '../lib1' })
          .addDeclarationImport({ declaration: declarations[1], from: '../lib2' });

        expect((ctrl as any).imports).to.matchSnapshot();
      });

      it('should add an import to an existing import index item', async () => {
        const ctrl = await provider(document);
        const declarations = [
          new ClassDeclaration('class1', true),
          new ClassDeclaration('class2', true),
        ];

        ctrl
          .addDeclarationImport({ declaration: declarations[0], from: '../lib1' })
          .addDeclarationImport({ declaration: declarations[1], from: '../lib1' });

        expect((ctrl as any).imports).to.matchSnapshot();
      });

      it('should not add the same specifier twice', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('class', true);

        ctrl.addDeclarationImport({ declaration, from: '../lib' });

        expect((ctrl as any).imports).to.matchSnapshot();

        ctrl.addDeclarationImport({ declaration, from: '../lib' });

        expect((ctrl as any).imports).to.matchSnapshot();
      });

      it('should add a normal import to a group', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('class', true);

        expect((ctrl as any).importGroups[2]).to.matchSnapshot();

        ctrl.addDeclarationImport({ declaration, from: '../lib' });

        expect((ctrl as any).importGroups[2]).to.matchSnapshot();
      });

      it('should add an import to an existing import group', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('class', true);
        const declaration2 = new ClassDeclaration('class2', true);

        expect((ctrl as any).importGroups[2]).to.matchSnapshot();

        ctrl
          .addDeclarationImport({ declaration, from: '../lib1' })
          .addDeclarationImport({ declaration: declaration2, from: '../lib1' });

        expect((ctrl as any).importGroups[2]).to.matchSnapshot();
      });

    });

    describe('organizeImports()', () => {

      it('should remove an unused import');

      it('should remove an unused specifier');

      it('should not remove an excluded library');

      it('should merge two same libraries into one import');

    });

    describe('commit()', () => {

      it('should not touch anything if nothing changed', async () => {
        const ctrl = await provider(document);

        await window.activeTextEditor!.edit((builder) => {
          builder.replace(
            document.lineAt(0).rangeIncludingLineBreak,
            `import {Class1} from '../resourceIndex';`,
          );
        });

        expect(await ctrl.commit()).to.be.true;
        expect(document.lineAt(0).text).to.equal(`import {Class1} from '../resourceIndex';`);
      });

      it('should add a single new import to the document top', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('importedclass', true);
        ctrl.addDeclarationImport({ declaration, from: '../imp' });
        await ctrl.commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should add two new imports to the document top', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('newClass', true);
        const declaration2 = new ClassDeclaration('secondnewClass', true);

        ctrl
          .addDeclarationImport({ declaration, from: '../lib' })
          .addDeclarationImport({ declaration: declaration2, from: '../not-same-lib' });
        await ctrl.commit();

        expect(getDocumentText(document, 0, 1)).to.matchSnapshot();
      });

      it('should add three new imports to the document top', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('newClass', true);
        const declaration2 = new ClassDeclaration('secondnewClass', true);
        const declaration3 = new ClassDeclaration('thirdnewClass', true);

        ctrl
          .addDeclarationImport({ declaration, from: '../lib' })
          .addDeclarationImport({ declaration: declaration2, from: '../not-same-lib' })
          .addDeclarationImport({ declaration: declaration3, from: '../not-same-lib-again' });
        await ctrl.commit();

        expect(getDocumentText(document, 0, 2)).to.matchSnapshot();
      });

      it('should add a single new module import to the document top', async () => {
        const ctrl = await provider(document);
        const declaration = new ModuleDeclaration('newModule');
        ctrl.addDeclarationImport({ declaration, from: 'new-module' });
        await ctrl.commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should add a single default import to the document top', async () => {
        const ctrl = await provider(document);
        const declaration = new DefaultDeclaration('defaultDeclaration', new File('file', rootPath, 1, 2));
        ctrl.addDeclarationImport({ declaration, from: '../lib' });
        await ctrl.commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should add a specifier to an existing import', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('class2', true);
        ctrl.addDeclarationImport({ declaration, from: '/server/indices' });
        await ctrl.commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should add multiple specifier to an existing import', async () => {
        const ctrl = await provider(document);
        const declaration = new ClassDeclaration('class2', true);
        const declaration2 = new ClassDeclaration('class3', true);

        await ctrl
          .addDeclarationImport({ declaration, from: '/server/indices' })
          .addDeclarationImport({ declaration: declaration2, from: '/server/indices' })
          .commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should add a specifier with a default (first) and a normal (second) import to the doc', async () => {
        const ctrl = await provider(document);
        const def = new DefaultDeclaration('defaultExport', new File(file.fsPath, rootPath, 1, 2));
        const dec = new ClassDeclaration('defaultClass', true);

        await ctrl
          .addDeclarationImport({ declaration: def, from: '/lib' })
          .addDeclarationImport({ declaration: dec, from: '/lib' })
          .commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should add a specifier to an import and a new import', async () => {
        const ctrl = await provider(document);
        const declaration1 = new ClassDeclaration('class2', true);
        const declaration2 = new ClassDeclaration('class3', true);

        await ctrl
          .addDeclarationImport({ declaration: declaration1, from: '/server/indices' })
          .addDeclarationImport({ declaration: declaration2, from: '/server/not-indices' })
          .commit();

        expect(getDocumentText(document, 0, 1)).to.matchSnapshot();
      });

      it('should convert a default import when a normal specifier is added', async () => {
        const ctrl = await provider(document);
        const def = new DefaultDeclaration('defaultExport', new File(file.fsPath, rootPath, 1, 2));
        const dec = new ClassDeclaration('defaultClass', true);

        await ctrl.addDeclarationImport({ declaration: def, from: '/server/indices' }).commit();

        expect(document.lineAt(0).text).to.matchSnapshot();

        await ctrl.addDeclarationImport({ declaration: dec, from: '/server/indices' }).commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should convert a normal import when a default specifier is added', async () => {
        const ctrl = await provider(document);
        const def = new DefaultDeclaration('defaultExport', new File(file.fsPath, rootPath, 1, 2));
        const dec = new ClassDeclaration('defaultClass', true);

        await ctrl.addDeclarationImport({ declaration: dec, from: '/server/indices' }).commit();

        expect(document.lineAt(0).text).to.matchSnapshot();

        await ctrl.addDeclarationImport({ declaration: def, from: '/server/indices' }).commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

      it('should render the optimized import', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.insert(new Position(5, 0), 'const foobar = new Class2();\n');
        });

        const ctrl = await provider(document);
        const declaration1 = new ClassDeclaration('Class2', true);
        const declaration2 = new ClassDeclaration('SomeOtherClass', true);

        await ctrl
          .addDeclarationImport({ declaration: declaration1, from: '/server/indices' })
          .addDeclarationImport({ declaration: declaration2, from: '/server/not-indices' })
          .commit();

        expect(getDocumentText(document, 0, 1)).to.matchSnapshot();

        await ctrl.organizeImports().commit();

        expect(getDocumentText(document, 0, 1)).to.matchSnapshot();
      });

      it('should render sorted imports when optimizing', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.insert(
            new Position(0, 0),
            `import { MultiExportClass } from '../server/indices/defaultExport/multiExport';\n`,
          );
          builder.insert(new Position(5, 0), 'const foobar = new MultiExportClass();\n');
        });
        const ctrl = await provider(document);

        await ctrl.organizeImports().commit();

        expect(getDocumentText(document, 0, 1)).to.matchSnapshot();
      });

      it('should render sorted specifiers when optimizing', async () => {
        await window.activeTextEditor!.edit((builder) => {
          builder.insert(new Position(0, 9), 'Class2, ');
          builder.insert(new Position(5, 0), 'const foobar = new Class2();\n');
        });
        const ctrl = await provider(document);

        await ctrl.organizeImports().commit();

        expect(document.lineAt(0).text).to.matchSnapshot();
      });

    });

  });

});
