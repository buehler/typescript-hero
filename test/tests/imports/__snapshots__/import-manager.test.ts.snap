// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportManager Typescript Constructor (via provider) should add an import proxy for a default import 1`] = `
Array [
  NamedImport {
    "defaultAlias": "myDefaultExportedFunction",
    "end": 84,
    "libraryName": "../defaultExport/lateDefaultExportedElement",
    "specifiers": Array [],
    "start": 0,
  },
]
`;

exports[`ImportManager Typescript Constructor (via provider) should add an import proxy for a named import 1`] = `
Array [
  NamedImport {
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
]
`;

exports[`ImportManager Typescript Constructor (via provider) should add multiple import proxies 1`] = `
Array [
  NamedImport {
    "defaultAlias": "myDefaultExportedFunction",
    "end": 84,
    "libraryName": "../defaultExport/lateDefaultExportedElement",
    "specifiers": Array [],
    "start": 0,
  },
  NamedImport {
    "end": 131,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 85,
  },
]
`;

exports[`ImportManager Typescript Constructor (via provider) should not add a proxy for a namespace import 1`] = `
Array [
  NamespaceImport {
    "alias": "bodyParser",
    "end": 42,
    "libraryName": "body-parser",
    "start": 0,
  },
]
`;

exports[`ImportManager Typescript Constructor (via provider) should not add a proxy for a string import 1`] = `
Array [
  StringImport {
    "end": 21,
    "libraryName": "body-parser",
    "start": 0,
  },
]
`;

exports[`ImportManager Typescript Constructor (via provider) should not add a proxy for an external import 1`] = `
Array [
  ExternalModuleImport {
    "alias": "bodyParser",
    "end": 43,
    "libraryName": "body-parser",
    "start": 0,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should add a default import to the import index. 1`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamedImport {
    "defaultAlias": "defaultdeclaration",
    "end": undefined,
    "libraryName": "../lib",
    "specifiers": Array [],
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should add a module import to the import index 1`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamespaceImport {
    "alias": "module",
    "end": undefined,
    "libraryName": "my-module",
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should add a normal import to a group 1`] = `
KeywordImportGroup {
  "imports": Array [
    NamedImport {
      "defaultAlias": undefined,
      "end": 46,
      "libraryName": "../../server/indices",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "Class1",
        },
      ],
      "start": 0,
    },
  ],
  "keyword": 2,
  "order": "asc",
}
`;

exports[`ImportManager Typescript addDeclarationImport() should add a normal import to a group 2`] = `
KeywordImportGroup {
  "imports": Array [
    NamedImport {
      "defaultAlias": undefined,
      "end": 46,
      "libraryName": "../../server/indices",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "Class1",
        },
      ],
      "start": 0,
    },
    NamedImport {
      "end": undefined,
      "libraryName": "../lib",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "class",
        },
      ],
      "start": undefined,
    },
  ],
  "keyword": 2,
  "order": "asc",
}
`;

exports[`ImportManager Typescript addDeclarationImport() should add a normal import to the document 1`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "classdeclaration",
      },
    ],
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should add an import to an existing import group 1`] = `
KeywordImportGroup {
  "imports": Array [
    NamedImport {
      "defaultAlias": undefined,
      "end": 46,
      "libraryName": "../../server/indices",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "Class1",
        },
      ],
      "start": 0,
    },
  ],
  "keyword": 2,
  "order": "asc",
}
`;

exports[`ImportManager Typescript addDeclarationImport() should add an import to an existing import group 2`] = `
KeywordImportGroup {
  "imports": Array [
    NamedImport {
      "defaultAlias": undefined,
      "end": 46,
      "libraryName": "../../server/indices",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "Class1",
        },
      ],
      "start": 0,
    },
    NamedImport {
      "end": undefined,
      "libraryName": "../lib1",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "class",
        },
      ],
      "start": undefined,
    },
    NamedImport {
      "end": undefined,
      "libraryName": "../lib1",
      "specifiers": Array [
        SymbolSpecifier {
          "alias": undefined,
          "specifier": "class2",
        },
      ],
      "start": undefined,
    },
  ],
  "keyword": 2,
  "order": "asc",
}
`;

exports[`ImportManager Typescript addDeclarationImport() should add an import to an existing import index item 1`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib1",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class1",
      },
    ],
    "start": undefined,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib1",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class2",
      },
    ],
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should add multiple imports to the import index 1`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib1",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class1",
      },
    ],
    "start": undefined,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib2",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class2",
      },
    ],
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should not add the same specifier twice 1`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class",
      },
    ],
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript addDeclarationImport() should not add the same specifier twice 2`] = `
Array [
  NamedImport {
    "defaultAlias": undefined,
    "end": 46,
    "libraryName": "../../server/indices",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "Class1",
      },
    ],
    "start": 0,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class",
      },
    ],
    "start": undefined,
  },
  NamedImport {
    "end": undefined,
    "libraryName": "../lib",
    "specifiers": Array [
      SymbolSpecifier {
        "alias": undefined,
        "specifier": "class",
      },
    ],
    "start": undefined,
  },
]
`;

exports[`ImportManager Typescript commit() should add a single new import to the document top 1`] = `"import { importedclass } from '../imp';"`;
